### PR TITLE
Testing ldl_factor_ex function but having issues generating hermitian data using hypothesis for testing 

### PR DESCRIPTION
**Pull request by keneccc**
wrote a function the function seems to be working. would love some feedback, I am not too sure if I was able to return the matrix LD in the correct form.

### CODE REVIEW

**Comment by sherry30** [comment]
Hi,

It would be better to create a test function and see if its working alright or fails anywhere.
Thanks

**Comment by keneccc** [comment]
Thank you! 

Heres the function being tested would love some feedback or any thoughts on how to update.

`def ldl_factor_ex(A, hermitian=False, check_errors=False):
    if hermitian:
        A = A.conj()

    n = A.shape[0]
    LD = torch.zeros_like(A)
    pivots = torch.zeros(n, dtype=torch.int32)
    info = torch.zeros(1, dtype=torch.int32)

    for j in range(n):
        for i in range(j):
            LD[j, i] = A[j, i]
            for k in range(i):
                LD[j, i] -= LD[j, k] * LD[i, k] * LD[k, k]

            LD[j, i] /= LD[i, i]

        LD[j, j] = A[j, j]
        for k in range(j):
            LD[j, j] -= LD[j, k] * LD[j, k] * LD[k, k]

        pivots[j] = j

    if check_errors:
        for j in range(n):
            if LD[j, j] == 0:
                info[0] = j + 1
                break

    if check_errors and info.item() != 0:
        raise RuntimeError(f"Zero diagonal element at index {info.item()} in LD factor. Division by zero may occur.")

    return LD, pivots, info


A = torch.randn(3, 3)
A = A @ A.mT # make symmetric


L,pivots_1, info_1 = ldl_factor_ex(A)`

**Comment by keneccc** [comment]
Here is the full function I do not think it sent correctly the first time
`import torch

def ldl_factor_ex(A, hermitian=False, check_errors=False):
    if hermitian:
        A = A.conj()

    n = A.shape[0]
    LD = torch.zeros_like(A)
    pivots = torch.zeros(n, dtype=torch.int32)
    info = torch.zeros(1, dtype=torch.int32)

    for j in range(n):
        for i in range(j):
            LD[j, i] = A[j, i]
            for k in range(i):
                LD[j, i] -= LD[j, k] * LD[i, k] * LD[k, k]

            LD[j, i] /= LD[i, i]

        LD[j, j] = A[j, j]
        for k in range(j):
            LD[j, j] -= LD[j, k] * LD[j, k] * LD[k, k]

        pivots[j] = j

    if check_errors:
        for j in range(n):
            if LD[j, j] == 0:
                info[0] = j + 1
                break

    if check_errors and info.item() != 0:
        raise RuntimeError(f"Zero diagonal element at index {info.item()} in LD factor. Division by zero may occur.")

    return LD, pivots, info


A = torch.randn(3, 3)
A = A @ A.mT # make symmetric

L,pivots_1, info_1 = ldl_factor_ex(A)


print(pivots_1)

**Comment by keneccc** [comment]
`
import torch

def ldl_factor_ex(A, hermitian=False, check_errors=False):
    if hermitian:
        A = A.conj()

    n = A.shape[0]
    LD = torch.zeros_like(A)
    pivots = torch.zeros(n, dtype=torch.int32)
    info = torch.zeros(1, dtype=torch.int32)

    for j in range(n):
        for i in range(j):
            LD[j, i] = A[j, i]
            for k in range(i):
                LD[j, i] -= LD[j, k] * LD[i, k] * LD[k, k]

            LD[j, i] /= LD[i, i]

        LD[j, j] = A[j, j]
        for k in range(j):
            LD[j, j] -= LD[j, k] * LD[j, k] * LD[k, k]

        pivots[j] = j

    if check_errors:
        for j in range(n):
            if LD[j, j] == 0:
                info[0] = j + 1
                break

    if check_errors and info.item() != 0:
        raise RuntimeError(f"Zero diagonal element at index {info.item()} in LD factor. Division by zero may occur.")

    return LD, pivots, info


A = torch.randn(3, 3)
A = A @ A.mT # make symmetric

L,pivots_1, info_1 = ldl_factor_ex(A)


print(pivots_1)
`

**Comment by sherry30** [comment]
HI,

You have to write the test using hypothesis. You can follow this guide [here](https://github.com/unifyai/ivy/blob/b708369a52165d95bed435956b5a9dc0625d4d65/docs/overview/deep_dive/ivy_frontends_tests.rst). This will test your function more thoroughly.

**Comment by keneccc** [comment]
ImportError: cannot import name 'config' from 'ivy_tests' (C:\Users\kenec\Documents\Miniconda\envs\ivy_dev\lib\site-packages\ivy_tests\__init__.py)
(ivy_dev) 

Hello I am trying to run the tests locally but I keep getting this error.

**Comment by keneccc** [comment]
```
@handle_frontend_test(
    fn_tree="torch.linalg.ldl_factor_ex",
    dtype_and_input=helpers.dtype_and_values(
    available_dtypes=helpers.get_dtypes("float")),
)
def test_torch_ldl_factor_ex(
    *,
    dtype_and_input,
    frontend,
    fn_tree,
    on_device,
    test_flags,
):
    input_dtype, x, N = dtype_and_input
    test_flags.num_positional_args = 1
    helpers.test_frontend_function(
        input_dtypes=input_dtype,
        frontend=frontend,
        fn_tree=fn_tree,
        on_device=on_device,
        test_flags=test_flags,
        x=x[0],
        N=N,
```

I wrote this function to test it

**Comment by sherry30** [comment]
Hmm seems like some kind of an environment issue to me. I'm not too familiar with miniconda setup, maybe @zaeemansari70 can comment on this.

Can you also push this code to your PR. Maybe there is a problem in the code. It'll be easier for me to look into it if its on your PR

**Comment by zaeemansari70** [comment]
@keneccc ,  installing Jax and Jaxlib should solve the issue 
For windows users:
```
pip install "jax[cpu]===0.4.9" -f https://whls.blob.core.windows.net/unstable/index.html --use-deprecated legacy-resolver
```

**Comment by keneccc** [comment]
I am also getting a key error with the function name so : KeyError: 'ldl_factor_ex'. Am I making any serious errors that could be causing this ?

**Comment by sherry30** [comment]
You have to write the test in test_frontends/test_torch/test_linalg.py file.
Can you also update your PR with the test function. would be easier for me to debug that way.

Thanks 😄

File: ivy_tests/test_ivy/test_frontends/test_torch/test_linalg.py [hunk 1]
```diff
@@ -1252,3 +1252,28 @@ def test_torch_multi_dot(
...
+)
+def test_torch_ldl_factor_ex(
+    *,
+    dtype_and_input,
+    frontend,
+    fn_tree,
+    on_device,
+    test_flags,
+):
+    input_dtype, x, N = dtype_and_input
```
Comment (sherry30) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_linalg.py:1269]: `dtype_and_input` only returns `input_dtype` and `x` not `N`

File: ivy_tests/test_ivy/test_frontends/test_torch/test_linalg.py [hunk 2]
```diff
@@ -1252,3 +1252,28 @@ def test_torch_multi_dot(
...
+    test_flags.num_positional_args = 1
+    helpers.test_frontend_function(
+        input_dtypes=input_dtype,
+        frontend=frontend,
+        fn_tree=fn_tree,
+        on_device=on_device,
+        test_flags=test_flags,
+        x=x[0],
+        N=N,
```
Comment (sherry30) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_linalg.py:1278]: the frontend function doesn't accept any argument `N`.
It accepts `hermitian`, you should generate and send this.

**Review by sherry30 — CHANGES_REQUESTED** [review]
Few changes requested

File: ivy_tests/test_ivy/test_frontends/test_torch/test_linalg.py (hunk 2 above)
Comment (keneccc) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_linalg.py]: I have made the requested adjustments and updated the files.

File: ivy_tests/test_ivy/test_frontends/test_torch/test_linalg.py [hunk 3]
```diff
@@ -1252,3 +1252,45 @@ def test_torch_multi_dot(
...
+    fn_tree="torch.linalg.ldl_factor_ex",
+    dtype_and_input=helpers.dtype_and_values(
+    available_dtypes=helpers.get_dtypes("float")),
+)
+import numpy as np
+#Size of parameter has been set as 3 and could be altered to alter results 
+def generate_hermitian_matrix():
+    size = 3  # Size is predetermined as 3x3
+    # Generate a random complex matrix
+    matrix = np.random.rand(size, size) + 1j * np.random.rand(size, size)
```
Comment (sherry30) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_linalg.py:1266]: We're not supposed to generate data using np random, we have to use hypothesis, feel free to read the docs to get an idea how that should be done.

File: ivy_tests/test_ivy/test_frontends/test_torch/test_linalg.py [hunk 4]
```diff
@@ -1252,3 +1252,45 @@ def test_torch_multi_dot(
         test_values=True,
         tensors=x,
     )
+
+@handle_frontend_test(
```
Comment (sherry30) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_linalg.py:1256]: This decorator should be on the test function.

**Review by sherry30 — CHANGES_REQUESTED** [review]
Hi,
Sorry for the delayed review

Requested a few changes.

Thanks

**Comment by keneccc** [comment]
Thank you for the feedback I have made the additonal adjustments

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by keneccc** [comment]
I have been stuck on testing this function and would love perhaos further clarification on utilising hypothesis to test the function. Am i supposed to test the original function using hypothesis or was it regarding simply generating the hermitian test using hypothesis

**Pull request closed without merging**